File: Python/103.Binary Tree Zigzag Level Order Traversal.py
#按照一条龙的顺序返回一棵二叉树
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def levelOrder(self, root):
        self.result = []
        self.inorder(root,0)
        for i in range(len(self.result)):           #把奇数行倒过来
            if i%2==1:
                self.result[i] = self.result[i][::-1]
        return self.result

    def inorder(self,p,level):
        if p!=None:
            if len(self.result)<=level:
                self.result.append([p.val])
            else:
                self.result[level].append(p.val)
            self.inorder(p.left,level+1) if p.left!=None else None
            self.inorder(p.right,level+1) if p.right!=None else None



t1 = TreeNode(1)
t2 = TreeNode(2)
t3 = TreeNode(3)
t4 = TreeNode(4)
t5 = TreeNode(5)
t6 = TreeNode(6)
t7 = TreeNode(7)
t1.left = t2
t1.right = t3
t2.left = t4
t2.right = t6
t3.right = t5
t3.left = t7


s = Solution()
print(s.levelOrder(t1))

File: C++/Tree/103.Binary Tree Zigzag Level Order Traversal.cpp
/*
按照zigzan的方式分层返回二叉树结点的值。
*/
/*
思路：分层返回结点时，奇数层的结点从前面插入。
 */
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
public:
	vector<vector<int>> result;
public:
	vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if(root==NULL)	return result;
        helpLevelOrder(root,0);
        return result;
    }
    void helpLevelOrder(TreeNode* cur,int level)
    {
    	if(cur!=NULL)
    	{
    		if(result.size()<=level)	result.push_back(vector<int> (1,cur->val));
    		else
    		{
    			if(level & 1) result[level].insert(result[level].begin(),cur->val);
    			else 	result[level].push_back(cur->val);
    		}
    		helpLevelOrder(cur->left,level+1);
    		helpLevelOrder(cur->right,level+1);
    	}
    }
};
int main()
{
	Solution s;
	TreeNode t1(1),t2(2),t3(3);
	t1.right = &t2;
	t2.left = &t3;
	vector<vector<int>> out = s.zigzagLevelOrder(&t1);
	for(auto vec:out)
	{
		for(auto n:vec)
			cout<<n;
		cout<<endl;
	}
}

